homeassistant/components/refoss/sensor.py

```diff
@@ -0,0 +1,194 @@
+
+    @callback
+    def init_device(coordinator):
+        """Register the device."""
+        device = coordinator.device
+
+        if not isinstance(device, ElectricityXMix):
+            return
+        descriptions = SENSORS.get(device.device_type)
+        new_entities = []
+        for channel in device.channels:
+            for description in descriptions:
+                entity = RefossSensor(
+                    coordinator=coordinator,
+                    channel=channel,
+                    description=description,
+                )
+                new_entities.append(entity)
+
+        async_add_entities(new_entities)
+
+    for coordinator in hass.data[DOMAIN][COORDINATORS]:
+        init_device(coordinator)
+
+    config_entry.async_on_unload(
+        async_dispatcher_connect(hass, DISPATCH_DEVICE_DISCOVERED, init_device)
+    )
+
+
+class RefossSensor(RefossEntity, SensorEntity):
+    """Refoss Sensor Device."""
+
+    entity_description: RefossSensorEntityDescription
+    _uom: UnitOfMeasurement | None = None
+
+    def __init__(
+        self,
+        coordinator: RefossDataUpdateCoordinator,
+        channel: int,
+        description: RefossSensorEntityDescription,
+    ) -> None:
+        """Init Refoss sensor."""
+        super().__init__(coordinator, channel)
+        self.entity_description = description
+        self._attr_unique_id = f"{super().unique_id}{description.key}"
+        device_type = coordinator.device.device_type
+        self._attr_name = (
+            f"{CHANNEL_DISPLAY_NAME[device_type][channel]} {description.name}"
+        )
+
+        if self.device_class is not None:
+            if (
+                self.native_unit_of_measurement is None
+                or self.device_class not in DEVICE_CLASS_UNITS
+            ):
+                self._attr_device_class = None
+                return
+
+            uoms = DEVICE_CLASS_UNITS[self.device_class]
+            self._uom = uoms.get(self.native_unit_of_measurement) or uoms.get(
+                self.native_unit_of_measurement.lower()
+            )
+            if self._uom is None:
+                self._attr_device_class = None
+                return
+            self._attr_native_unit_of_measurement = (
+                self._uom.conversion_unit or self._uom.unit
+            )
+
+    @property
+    def native_value(self) -> StateType:
+        """Return the native value."""
+        value = self.coordinator.device.get_value(
+            self.channel_id, self.entity_description.subkey
+        )
+        if value is None:
+            return None
+        if (self.entity_description.key == ENERGY and value < 0) or (
+            self.entity_description.key == ENERGY_RETURNED and value > 0
+        ):
+            value = 0
+
+        if self._uom and self._uom.conversion_fn is not None:
+            return self._uom.conversion_fn(value)
+
+        if isinstance(value, float):
+            return f"{value:.2f}"
```

Comment:
Numeric values should not converted to strings. The user can change how many digits are shown in the UI and therefore we should always return the value as number